File: compiler/runtime/builtins.c
#include <math.h>
#include <stdio.h>
#include "builtins.h"

Call builtin_times(Env env, Args args) {
    Number left = (Number)args_get(args, 0);
    Number right = (Number)args_get(args, 1);
    Closure k = (Closure)args_get(args, 2);
    Args next_args = create_args(1);
    args_set(next_args, 0, const_number(left->value * right->value));
    return create_call(k, next_args);
}

Call builtin_plus(Env env, Args args) {
    Number left = (Number)args_get(args, 0);
    Number right = (Number)args_get(args, 1);
    Closure k = (Closure)args_get(args, 2);
    Args next_args = create_args(1);
    args_set(next_args, 0, const_number(left->value + right->value));
    return create_call(k, next_args);
}

Call builtin_minus(Env env, Args args) {
    Number left = (Number)args_get(args, 0);
    Number right = (Number)args_get(args, 1);
    Closure k = (Closure)args_get(args, 2);
    Args next_args = create_args(1);
    args_set(next_args, 0, const_number(left->value - right->value));
    return create_call(k, next_args);
}

Call builtin_sqrt(Env env, Args args) {
    Number n = (Number)args_get(args, 0);
    Closure k = (Closure)args_get(args, 1);
    Args next_args = create_args(1);
    args_set(next_args, 0, const_number(sqrt(n->value)));
    return create_call(k, next_args);
}

Call builtin_printNumber(Env env, Args args) {
    Number n = (Number)args_get(args, 0);
    Closure k = (Closure)args_get(args, 1);
    printf("%f\n", n->value);
    return create_call(k, create_args(0));
}

Call builtin_isZeroP(Env env, Args args) {
    Number n = (Number)args_get(args, 0);
    Closure trueFn = (Closure)args_get(args, 1);
    Closure falseFn = (Closure)args_get(args, 2);
    if (n->value == 0) {
        return create_call(trueFn, create_args(0));
    } else {
        return create_call(falseFn, create_args(0));
    }
}

Call builtin_setTempo(Env env, Args args) {
    Number n = (Number)args_get(args, 0);
    Closure k = (Closure)args_get(args, 1);
    printf("setting tempo: %f\n", n->value);
    return create_call(k, create_args(0));
}

Call builtin_setBeat1(Env env, Args args) {
    Number n = (Number)args_get(args, 0);
    Closure k = (Closure)args_get(args, 1);
    printf("setting beat 1: %f\n", n->value);
    return create_call(k, create_args(0));
}

Call builtin_exit(Env env, Args args) {
    return NULL;
}

File: compiler/compiler/Builtins.hs
module Builtins where

import qualified Types.Types as T
import Types.Semantic

builtins =
    [ Builtin
        { name = "times"
        , code = "k = arg2; next_args = create_args(1); args_set(next_args, 0, const_number(arg0->value * arg1->value)); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Number, T.Function [T.Number]]
        , includes = []
        }
    , Builtin
        { name = "plus"
        , code = "k = arg2; next_args = create_args(1); args_set(next_args, 0, const_number(arg0->value + arg1->value)); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Number, T.Function [T.Number]]
        , includes = []
        }
    , Builtin
        { name = "minus"
        , code = "k = arg2; next_args = create_args(1); args_set(next_args, 0, const_number(arg0->value - arg1->value)); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Number, T.Function [T.Number]]
        , includes = []
        }
    , Builtin
        { name = "sqrt"
        , code = "k = arg1; next_args = create_args(1); args_set(next_args, 0, const_number(sqrt(arg0->value))); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Function [T.Number]]
        , includes = ["<math.h>"]
        }
    , Builtin
        { name = "printNumber"
        , code = "k = arg1; next_args = create_args(0); printf(\"%f\\n\", arg0->value); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Function []]
        , includes = ["<stdio.h>"]
        }
    , Builtin
        { name = "exit"
        , code = "return NULL;"
        , fnType = T.Function []
        , includes = []
        }
    , Builtin
        { name = "isZero?"
        , code = "next_args = create_args(0); if (arg0->value == 0) { k = arg1; } else { k = arg2; }; return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Function [T.Number], T.Function [T.Number]]
        , includes = []
        }
    , Builtin
        { name = "setTempo"
        , code = "k = arg1; next_args = create_args(0); printf(\"set tempo: %f\\n\", arg0->value); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Function []]
        , includes = ["<stdio.h>"]
        }
    , Builtin
        { name = "setBeat1"
        , code = "k = arg1; next_args = create_args(0); printf(\"set beat 1: %f\\n\", arg0->value); return create_call(k, next_args);"
        , fnType = T.Function [T.Number, T.Function []]
        , includes = ["<stdio.h>"]
        }
    ]

addBuiltins :: Program -> Program
addBuiltins (Program lets) = Program (newLets ++ lets)
    where
        newLets = map (\(b) -> Let (name b) (Function (Fn (fnType b) b))) builtins
